File: go-app/internal/wicketkinds/wicketkinds.go
// Package wicketkinds reads the vocabulary of wicket kinds and what each is to the
// scorecard.
//
// A wicket in a Cricsheet file is one of fourteen kinds, and they are not all the same
// thing: six are the bowler's (bowled, caught, caught and bowled, hit wicket, lbw,
// stumped), six are dismissals nobody bowls (run out, retired out, obstructing the field,
// handled the ball, hit the ball twice, timed out), and two are not dismissals at all -- a
// batter who retires hurt or retires not out may come back, and the scorecard does not
// count him as a wicket lost. Until IMPORT-06 was fixed the importer credited every kind
// to the bowler and counted every kind as a wicket lost.
//
// The vocabulary is committed data (configs/wicket_kinds.json), read by this package and
// by the rating pass (ml/xi/wicketkinds.py), so the two languages cannot carry different
// sets: the importer's bowling_data.wickets and the pass's wickets target are one rule.
// A kind that is not in the file is an error rather than a guess, because the only way a
// new kind arrives is Cricsheet adding one, and the right answer to that is a reviewed
// edit to the file.
package wicketkinds

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
)

// FileName is the vocabulary's name under configs/.
const FileName = "wicket_kinds.json"

// EnvVar overrides where the vocabulary is read from.
const EnvVar = "GO_APP_WICKET_KINDS"

// Class is what a kind of wicket is to the scorecard.
type Class int

const (
	// CreditedToBowler is a dismissal that goes into the bowler's wickets column.
	CreditedToBowler Class = iota + 1
	// DismissalNotCredited is a dismissal that is nobody's: the innings loses a wicket,
	// the bowler's figures do not move.
	DismissalNotCredited
	// NotOut is a batter who left the crease without being dismissed and may return.
	NotOut
)

// IsDismissal reports whether the innings lost a wicket.
func (c Class) IsDismissal() bool {
	return c == CreditedToBowler || c == DismissalNotCredited
}

// String names the class the way the vocabulary file does.
func (c Class) String() string {
	switch c {
	case CreditedToBowler:
		return "credited_to_bowler"
	case DismissalNotCredited:
		return "dismissal_not_credited"
	case NotOut:
		return "not_out"
	default:
		return fmt.Sprintf("Class(%d)", int(c))
	}
}

// Kind is one wicket kind as the vocabulary spells it, and its class.
type Kind struct {
	Name  string
	Class Class
}

// Vocabulary is the reviewed set of wicket kinds.
type Vocabulary struct {
	Version              string   `json:"version"`
	CreditedToBowler     []string `json:"credited_to_bowler"`
	DismissalNotCredited []string `json:"dismissal_not_credited"`
	NotOut               []string `json:"not_out"`
}

// Kind resolves the kind a file spells -- case and surrounding space aside -- to the
// vocabulary's spelling and class. The archive has always spelled every kind exactly as
// the file does, so the normalisation is defensive; the error is not.
func (v Vocabulary) Kind(raw string) (Kind, error) {
	name := normalise(raw)
	for class, names := range v.byClass() {
		for _, candidate := range names {
			if candidate == name {
				return Kind{Name: candidate, Class: class}, nil
			}
		}
	}
	return Kind{}, fmt.Errorf("wicket kind %q is not in the vocabulary (%s)", raw, FileName)
}

// Validate rejects a vocabulary that could classify a kind two ways or none.
func (v Vocabulary) Validate() error {
	seen := map[string]Class{}
	for class, names := range v.byClass() {
		if len(names) == 0 {
			return fmt.Errorf("the %s list is empty", class)
		}
		for _, name := range names {
			if name != normalise(name) || name == "" {
				return fmt.Errorf("%q is not in the vocabulary's own spelling (lower case, trimmed)", name)
			}
			if previous, dup := seen[name]; dup {
				return fmt.Errorf("%q is listed twice (%s and %s); a kind is one thing", name, previous, class)
			}
			seen[name] = class
		}
	}
	return nil
}

func (v Vocabulary) byClass() map[Class][]string {
	return map[Class][]string{
		CreditedToBowler:     v.CreditedToBowler,
		DismissalNotCredited: v.DismissalNotCredited,
		NotOut:               v.NotOut,
	}
}

func normalise(raw string) string {
	return strings.ToLower(strings.TrimSpace(raw))
}

// Load reads the vocabulary from GO_APP_WICKET_KINDS, or from configs/ searched upwards
// from the working directory, matching how config.json and the team lineage are found.
//
// Unlike the lineage, a missing file is an error: nothing can say what a run out is to the
// bowler without it, and an import that guessed would be the defect this file exists to
// close.
func Load() (Vocabulary, error) {
	path := os.Getenv(EnvVar)
	if path == "" {
		path = find()
	}
	if path == "" {
		slog.Error("wicketkinds: no vocabulary found",
			slog.String("looked_for", filepath.Join("configs", FileName)),
			slog.String("env", EnvVar))
		return Vocabulary{}, fmt.Errorf("no %s found under configs/ (set %s)", FileName, EnvVar)
	}
	return LoadFile(path)
}

// LoadFile reads and validates one vocabulary file.
func LoadFile(path string) (Vocabulary, error) {
	b, err := os.ReadFile(path) // #nosec G304 -- operator-supplied config path
	if err != nil {
		slog.Error("wicketkinds: read failed", slog.String("path", path), slog.Any("err", err))
		return Vocabulary{}, fmt.Errorf("read %s: %w", path, err)
	}
	var v Vocabulary
	if err := json.Unmarshal(b, &v); err != nil {
		slog.Error("wicketkinds: parse failed", slog.String("path", path), slog.Any("err", err))
		return Vocabulary{}, fmt.Errorf("parse %s: %w", path, err)
	}
	if err := v.Validate(); err != nil {
		slog.Error("wicketkinds: vocabulary is not usable", slog.String("path", path), slog.Any("err", err))
		return Vocabulary{}, fmt.Errorf("%s: %w", path, err)
	}
	slog.Info("wicketkinds: loaded",
		slog.String("path", path),
		slog.Int("kinds", len(v.CreditedToBowler)+len(v.DismissalNotCredited)+len(v.NotOut)))
	return v, nil
}

func find() string {
	for _, dir := range []string{".", "..", filepath.Join("..", ".."), filepath.Join("..", "..", "..")} {
		p := filepath.Join(dir, "configs", FileName)
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	return ""
}

File: ml-service/ml/xi/wicketkinds.py
"""The vocabulary of wicket kinds, and what each one is to the scorecard (IMPORT-06).

A wicket in a Cricsheet file is one of fourteen kinds, and they are not all the same thing:
six are the bowler's (bowled, caught, caught and bowled, hit wicket, lbw, stumped), six are
dismissals nobody bowls (run out, retired out, obstructing the field, handled the ball,
hit the ball twice, timed out), and two are not dismissals at all -- a batter who retires
hurt or retires not out may come back. Until IMPORT-06 was fixed this pass counted a
retired-hurt batter as a dismissal, and the go-app importer credited every kind to the
bowler.

``configs/wicket_kinds.json`` is the one list. The importer reads it (``internal/wicketkinds``)
for ``bowling_data.wickets`` and ``match_inning.wickets_lost``; this module reads it for the
``wickets`` and ``dismissals`` targets and the innings' wicket counts, on both rating
sources. Both sides reading one file is the point: a set carried in each language would be
two sets the day one of them was edited. A kind the file does not name is an error rather
than a guess, because the only way a new kind arrives is Cricsheet adding one, and the
right answer to that is a reviewed edit to the file.
"""

from __future__ import annotations

import json
import logging
import os
from dataclasses import dataclass
from functools import lru_cache
from typing import FrozenSet, Optional, Sequence

logger = logging.getLogger(__name__)

FILE_NAME = "wicket_kinds.json"
ENV_VAR = "GO_APP_WICKET_KINDS"

#: The one kind the keeper flag reads: a stumping is the bowler's wicket and the keeper's credit.
STUMPED = "stumped"


def normalise(kind: str) -> str:
    """A kind as the vocabulary spells it: lower case, trimmed. The archive has always
    spelled every kind exactly so; this is defensive, the same way the importer is."""
    return kind.strip().lower()


@dataclass(frozen=True)
class Vocabulary:
    """The reviewed set of wicket kinds, by what the scorecard makes of each."""

    credited_to_bowler: FrozenSet[str]
    dismissal_not_credited: FrozenSet[str]
    not_out: FrozenSet[str]

    def __post_init__(self) -> None:
        classes = {
            "credited_to_bowler": self.credited_to_bowler,
            "dismissal_not_credited": self.dismissal_not_credited,
            "not_out": self.not_out,
        }
        seen: dict = {}
        for name, kinds in classes.items():
            if not kinds:
                raise ValueError(f"the {name} list is empty")
            for kind in kinds:
                if not kind or kind != normalise(kind):
                    raise ValueError(f"{kind!r} is not in the vocabulary's own spelling (lower case, trimmed)")
                if kind in seen:
                    raise ValueError(f"{kind!r} is listed twice ({seen[kind]} and {name}); a kind is one thing")
                seen[kind] = name

    def _known(self, kind: str) -> str:
        name = normalise(kind)
        if name not in self.credited_to_bowler and name not in self.dismissal_not_credited and name not in self.not_out:
            raise ValueError(f"wicket kind {kind!r} is not in the vocabulary ({FILE_NAME})")
        return name

    def is_dismissal(self, kind: str) -> bool:
        """Whether the innings lost a wicket: every kind but a retirement not out."""
        return self._known(kind) not in self.not_out

    def is_credited_to_bowler(self, kind: str) -> bool:
        """Whether the wicket goes into the bowler's column."""
        return self._known(kind) in self.credited_to_bowler

    def is_stumping(self, kind: str) -> bool:
        return self._known(kind) == STUMPED


def _search_paths() -> Sequence[str]:
    here = os.path.dirname(os.path.abspath(__file__))
    return (
        os.path.join(here, "..", "..", "..", "configs", FILE_NAME),
        os.path.join("configs", FILE_NAME),
        os.path.join("..", "configs", FILE_NAME),
    )


def load(path: Optional[str] = None) -> Vocabulary:
    """Read the vocabulary from ``path``, else from ``GO_APP_WICKET_KINDS``, else from
    ``configs/`` beside the repository.

    Unlike the team lineage, an absent file is an error: nothing can say what a run out is
    to the bowler without it, and a pass that guessed would be the defect the file exists
    to close. A file that does not parse or does not validate is an error for the same
    reason the lineage's is: it is committed data.
    """
    candidates = [path] if path else [os.environ.get(ENV_VAR) or "", *_search_paths()]
    for candidate in candidates:
        if candidate and os.path.exists(candidate):
            with open(candidate) as fh:
                raw = json.load(fh)
            vocabulary = Vocabulary(
                credited_to_bowler=frozenset(raw.get("credited_to_bowler") or []),
                dismissal_not_credited=frozenset(raw.get("dismissal_not_credited") or []),
                not_out=frozenset(raw.get("not_out") or []),
            )
            logger.info(
                "wicket kinds: %d kinds from %s",
                len(vocabulary.credited_to_bowler) + len(vocabulary.dismissal_not_credited) + len(vocabulary.not_out),
                candidate,
            )
            return vocabulary
    logger.error("wicket kinds: no %s found under configs/ (set %s)", FILE_NAME, ENV_VAR)
    raise FileNotFoundError(f"no {FILE_NAME} found under configs/ (set {ENV_VAR})")


@lru_cache(maxsize=1)
def vocabulary() -> Vocabulary:
    """The committed vocabulary, read once per process."""
    return load()
